virtcontainers/sandbox.go

```diff
@@ -1923,9 +1926,30 @@ func (s *Sandbox) calculateSandboxMemory() int64 {
 
 		if m := c.Resources.Memory; m != nil && m.Limit != nil {
 			memorySandbox += *m.Limit
+			s.Logger().WithField("memory limit", memorySandbox).Info("Memory Sandbox + Memory Limit ")
+		}
+
+		//Add hugepages memory
+		//HugepageLimit is uint64 - https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L242
```

Comment:
This relates to https://github.com/kata-containers/runtime/pull/3109#discussion_r551415801. However, I'm looking at the code again and wondering why we cannot simply change `calculateSandboxMemory()` to return `uint64` rather than `int64`?

Reply:
> This relates to [#3109 (comment)](https://github.com/kata-containers/runtime/pull/3109#discussion_r551415801). However, I'm looking at the code again and wondering why we cannot simply change `calculateSandboxMemory()` to return `uint64` rather than `int64`?

Memory limit is in int64 whereas hugepage limit is in unit64 .  One of the entities will need conversion.
https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L298
https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L242

Reply:
Can't we do the following on top of your PR?:

```diff
diff --git a/virtcontainers/sandbox.go b/virtcontainers/sandbox.go
index 9f0a2f45..c87d4c4f 100644
--- a/virtcontainers/sandbox.go
+++ b/virtcontainers/sandbox.go
@@ -1877,7 +1877,7 @@ func (s *Sandbox) updateResources() error {
 		return err
 	}
 	// Add default / rsvd memory for sandbox.
-	sandboxMemoryByte += int64(s.hypervisor.hypervisorConfig().MemorySize) << utils.MibToBytesShift
+	sandboxMemoryByte += uint64(s.hypervisor.hypervisorConfig().MemorySize) << utils.MibToBytesShift
 
 	// Update VCPUs
 	s.Logger().WithField("cpus-sandbox", sandboxVCPUs).Debugf("Request to hypervisor to update vCPUs")
@@ -1897,7 +1897,9 @@ func (s *Sandbox) updateResources() error {
 
 	// Update Memory
 	s.Logger().WithField("memory-sandbox-size-byte", sandboxMemoryByte).Debugf("Request to hypervisor to update memory")
-	newMemory, updatedMemoryDevice, err := s.hypervisor.resizeMemory(uint32(sandboxMemoryByte>>utils.MibToBytesShift), s.state.GuestMemoryBlockSizeMB, s.state.GuestMemoryHotplugProbe)
+	newMemoryMB := uint32(sandboxMemoryByte >> utils.MibToBytesShift)
+
+	newMemory, updatedMemoryDevice, err := s.hypervisor.resizeMemory(newMemoryMB, s.state.GuestMemoryBlockSizeMB, s.state.GuestMemoryHotplugProbe)
 	if err != nil {
 		return err
 	}
@@ -1915,8 +1917,8 @@ func (s *Sandbox) updateResources() error {
 	return nil
 }
 
-func (s *Sandbox) calculateSandboxMemory() (int64, error) {
-	memorySandbox := int64(0)
+func (s *Sandbox) calculateSandboxMemory() (uint64, error) {
+	memorySandbox := uint64(0)
 	for _, c := range s.config.Containers {
 		// Do not hot add again non-running containers resources
 		if cont, ok := s.containers[c.ID]; ok && cont.state.State == types.StateStopped {
@@ -1924,8 +1926,8 @@ func (s *Sandbox) calculateSandboxMemory() (int64, error) {
 			continue
 		}
 
-		if m := c.Resources.Memory; m != nil && m.Limit != nil {
-			memorySandbox += *m.Limit
+		if m := c.Resources.Memory; m != nil && m.Limit != nil && *m.Limit > 0 {
+			memorySandbox += uint64(*m.Limit)
 			s.Logger().WithField("memory limit", memorySandbox).Info("Memory Sandbox + Memory Limit ")
 		}
 
@@ -1933,7 +1935,7 @@ func (s *Sandbox) calculateSandboxMemory() (int64, error) {
 		//HugepageLimit is uint64 - https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L242
 		for _, l := range c.Resources.HugepageLimits {
 			if l.Limit <= math.MaxInt64 {
-				memorySandbox += int64(l.Limit)
+				memorySandbox += uint64(l.Limit)
 			} else {
 				//Overflow
 				s.Logger().WithField("hugepages memory limit overflow", l.Limit).Warn("Memory Sandbox + Hugepages Memory Limit ")
```

Reply:
@jodh-intel the current iteration incorporates your above suggestion ..  ptal